uvicorn/protocols/websockets/websockets_impl.py

```diff
@@ -278,43 +282,83 @@ async def asgi_send(self, message: "ASGISendEvent") -> None:
+
+                elif message_type == "websocket.close":
+                    message = cast("WebSocketCloseEvent", message)
+                    self.logger.info(
+                        '%s - "WebSocket %s" 403',
+                        self.scope["client"],
+                        get_path_with_query_string(self.scope),
+                    )
+                    self.initial_response = (http.HTTPStatus.FORBIDDEN, [], b"")
+                    self.handshake_started_event.set()
+                    self.closed_event.set()
+
+                elif message_type == "websocket.http.response.start":
+                    message = cast("WebSocketResponseStartEvent", message)
+                    try:
+                        status = http.HTTPStatus(message["status"])
+                    except AttributeError:
+                        status = http.HTTPStatus.FORBIDDEN
+                    headers = [
                         (name.decode("latin-1"), value.decode("latin-1"))
-                        for name, value in message["headers"]
+                        for name, value in message.get("headers", [])
+                    ]
+                    self.initial_response = (status, headers, b"")
+                    self.response_body = []
+
+                else:
+                    msg = (
+                        "Expected ASGI message 'websocket.accept', 'websocket.close', "
+                        "or 'websocket.http.response.start' "
+                        "but got '%s'."
                     )
-                self.handshake_started_event.set()
-
-            elif message_type == "websocket.close":
-                message = cast("WebSocketCloseEvent", message)
-                self.logger.info(
-                    '%s - "WebSocket %s" 403',
-                    self.scope["client"],
-                    get_path_with_query_string(self.scope),
-                )
-                self.initial_response = (http.HTTPStatus.FORBIDDEN, [], b"")
-                self.handshake_started_event.set()
-                self.closed_event.set()
-
+                    raise RuntimeError(msg % message_type)
             else:
-                msg = (
-                    "Expected ASGI message 'websocket.accept' or 'websocket.close', "
-                    "but got '%s'."
-                )
-                raise RuntimeError(msg % message_type)
+                if message_type == "websocket.http.response.body":
+                    message = cast("WebSocketResponseBodyEvent", message)
+                    self.response_body.append(message["body"])
+
+                    if not message.get("more_body", False):
+                        self.logger.info(
+                            '%s - "WebSocket %s" %i',
+                            self.scope["client"],
+                            get_path_with_query_string(self.scope),
+                            self.initial_response[0],
+                        )
+
```

Comment:
Access logs should be sent on `start` messages.
```suggestion
```

Reply:
done